### PR TITLE
[Kernel Selection] Remove hard code of PADDLE_WITH_CUDA

### PR DESCRIPTION
**Pull request by jiahy0825**
### PR types
Function optimization

### PR changes
OPs

### Describe
Delete the hardcodes of `if defined(PADDLE_WITH_CUDA) || defined(PADDLE_WITH_HIP)` and move such logic outside `GetExpectedKernelType`.

Relevant ops: `affine_grid`, `affine_grid_grad`, `conv2d_transpose_grad`, `conv2d_transpose_grad_grad`, `conv3d_transpose_grad`, `depthwise_conv2d_transpose_grad`, `grid_sampler`, `grid_sampler_grad`, `pool2d`, `pool2d_grad`, `pool2d_double_grad`, `pool3d`, `pool3d_grad`, `softmax`, `softmax_grad` and `activation ops`.

### CODE REVIEW

File: paddle/fluid/imperative/execution_context.h [hunk 1]
```diff
@@ -103,7 +103,8 @@ class DygraphExecutionContext : public framework::ExecutionContext {
 
   bool HasAttr(const std::string& name) const override {
     if (attrs_.find(name) == attrs_.end()) {
-      return default_attrs_.find(name) != default_attrs_.end();
+      return &default_attrs_ != nullptr &&
```
Comment (YuanRisheng) [inline paddle/fluid/imperative/execution_context.h:106]: 这里为何需要判断Nullptr，default_attrs_是一个实例对象，应该不会出现地址是nullptr的情况吧

**Review by YuanRisheng — APPROVED** [review]

File: paddle/fluid/operators/activation_op.cc [hunk 2]
```diff
@@ -93,6 +93,14 @@ framework::OpKernelType GetKernelType(const framework::ExecutionContext& ctx,
...
   //   }
   // #endif
+
+  // NOTE(jiahongyu): Activation ops have attribute use_cudnn, but cudnn kernels
+  // are temporarily disabled. Therefore, cudnn kernel also needs to fallback to
+  // plain GPU kernel temporarily. When above codes are uncommented, below
+  // fallback codes can be deleted safely.
+  if (paddle::platform::is_gpu_place(ctx.GetPlace())) {
+    oper.SetDnnFallback(true);
+  }
```
Comment (zyfncg) [inline paddle/fluid/operators/activation_op.cc:103]: 通过Yaml生成的激活函数算子没有这段逻辑有的会执行到cudnn的kernel，可能需要留意下对模型是否有影响

**Review by zyfncg — APPROVED** [review]

File: paddle/fluid/imperative/execution_context.h (hunk 1 above)
Comment (jiahy0825) [reply paddle/fluid/imperative/execution_context.h:106]: https://xly.bce.baidu.com/paddlepaddle/paddle/newipipe/detail/7013430/job/19820813
以上述 CI 报错为例，可能出现空指针，经过分析，当 op 只有输入输出，没有 attr 时，可能会出现 `attrs_.size() == 0 && default_attrs_ == nullptr` 的情况。`assign`, `py_layer` 两个 op 都会出问题，并且报错出现在老动态图的调用流程中，解决这个问题还需要进一步调试。

According to the above PR-CI-Coverage link, default_attrs_ may be a nullptr. When the op only has inputs and outputs and the op has no attr, such a scenario may occur: `attrs_.size() == 0 && default_attrs_ == nullptr`. At least `assign` and `py_layer` have this bug, and this bug will occur in the calling process of the legacy dynamic graph.

File: paddle/fluid/operators/activation_op.cc (hunk 2 above)
Comment (jiahy0825) [reply paddle/fluid/operators/activation_op.cc:103]: 好的，激活函数算子都没有执行 cudnn kernel，一种更好的方式也许是直接删除激活函数中的 use_cudnn 属性，然而基于如下考虑，采用了 Fallback 的实现方式：
1. 激活函数算子删除 use_cudnn 属性后，需要同步删除所有激活函数单测中用到 use_cudnn 的代码，这部分工作量稍大
2. 目前激活函数只是暂时不支持 cudnn kernel，日后支持 cudnn kernel 后，还需要把 use_cudnn 属性添加回来
3. 极小概率下，推理模型可能会不兼容

Currently, all activation kernels are not calling the cudnn kernel, maybe it's a better way to delete `use_cudnn` attribute directly. However, based on the following consideration, I adopt the implementation of fallback:
1. If activation ops delete the `use_cudnn` attribute, we need to delete all `use_cudnn` attributes simultaneously in the activation unit tests. This involves a lot of hard work.
2. Activation cudnn kernels are temporarily disabled right now, if we support cudnn kernels in the future, we will need to restore the `use_cudnn` attribute back again.
3. Under extremely small probability, the inference model may have incompatibility problems.

**Review by chenwhql — APPROVED** [review]

File: paddle/fluid/imperative/execution_context.h (hunk 1 above)
Comment (jiahy0825) [reply paddle/fluid/imperative/execution_context.h:106]: Fix `default_attrs_ == nullptr`: [[Opt Kernel Selection] Opt CanMKLDNNBeUsed performance #47563](https://github.com/PaddlePaddle/Paddle/pull/47563)

**Pull request merged**